`rustc_scalable_vector(N)`

davidtwco opened the pull request:
Supercedes rust-lang/rust#118917.

Initial experimental implementation of rust-lang/rfcs#3838. Introduces a `rustc_scalable_vector(N)` attribute that can be applied to types with a single `[$ty]` field (for `u{16,32,64}`, `i{16,32,64}`, `f{32,64}`, `bool`). `rustc_scalable_vector` types are lowered to scalable vectors in the codegen backend.

As with any unstable feature, there will necessarily be follow-ups as we experiment and find cases that we've not considered or still need some logic to handle, but this aims to be a decent baseline to start from.

See rust-lang/rust#145052 for request for a lang experiment.

davidtwco committed:
attr: parse `rustc_scalable_vector(N)`

Extend parsing of `ReprOptions` with `rustc_scalable_vector(N)` which
optionally accepts a single literal integral value - the base multiple of
lanes that are in a scalable vector. Can only be applied to structs.

Co-authored-by: Jamie Cunliffe <Jamie.Cunliffe@arm.com>

davidtwco committed:
hir/trait_sel: prohibit scalable vectors in types

Extend well-formedness checking and HIR analysis to prohibit the use of
scalable vectors in structs, enums, unions, tuples and arrays. LLVM does
not support scalable vectors being members of other types, so these
restrictions are necessary.

Co-authored-by: Jamie Cunliffe <Jamie.Cunliffe@arm.com>

davidtwco committed:
mir_transform: prohibit scalable vectors in async

Scalable vectors cannot be members of ADTs and thus cannot be kept over
await points in async functions.

davidtwco committed:
mono: require target feature for scalable vectors

Scalable vector types only work with the relevant target features
enabled, so require this for any function with the types in its
signature.

davidtwco committed:
codegen: implement `repr(scalable)`

Introduces `BackendRepr::ScalableVector` corresponding to scalable
vector types annotated with `repr(scalable)` which lowers to a scalable
vector type in LLVM.

Co-authored-by: Jamie Cunliffe <Jamie.Cunliffe@arm.com>

davidtwco committed:
debuginfo: no spill `<vscale x N x i1>` for `N!=16`

LLVM doesn't handle stores on `<vscale x N x i1>` for `N != 16`, a type
used internally in SVE intrinsics. Spilling to the stack to create
debuginfo will cause errors during instruction selection. These types
that are an internal implementation detail to the intrinsic, so users
should never see them types and won't need any debuginfo.

Co-authored-by: Jamie Cunliffe <Jamie.Cunliffe@arm.com>

davidtwco commented:
I've changed this back to a draft and marked it as `S-waiting-on-author` while I rebase it and get the approval for an experimental implementation to go ahead.

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16266498150/job/45923468428) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45923468428) [(plain)](https://github.com/rust-lang/rust/commit/3ad08987db242a15bc7dfe3a5eac91597179feff/checks/45923468428/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] form_urlencoded test:false 0.072
error[E0027]: pattern does not mention field `scalable`
   --> src/tools/rust-analyzer/crates/hir-def/src/attr.rs:334:9
    |
334 |     let ReprOptions { int, align, pack, flags, field_shuffle_seed: _ } = this;
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing field `scalable`
    |
help: include the missing field in the pattern
    |
334 |     let ReprOptions { int, align, pack, flags, field_shuffle_seed: _, scalable } = this;
    |                                                                     ++++++++++
help: if you don't care about this missing field, you can explicitly ignore it
    |
334 |     let ReprOptions { int, align, pack, flags, field_shuffle_seed: _, scalable: _ } = this;
    |                                                                     +++++++++++++
help: or always ignore missing fields here
    |
334 |     let ReprOptions { int, align, pack, flags, field_shuffle_seed: _, .. } = this;
    |                                                                     ++++

[RUSTC-TIMING] thiserror_impl test:false 1.562
    Checking test-utils v0.0.0 (/checkout/src/tools/rust-analyzer/crates/test-utils)
[RUSTC-TIMING] test_utils test:false 0.234
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16266927500/job/45924745691) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45924745691) [(plain)](https://github.com/rust-lang/rust/commit/5c92874ae20b71f01ace604511c3878914746c0b/checks/45924745691/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 48)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 51)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 68)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
---- [ui] tests/ui/simd/scalable/copy-clone.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/simd/scalable/copy-clone.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=i686-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/simd/scalable/copy-clone" "-A" "unused" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/i686-unknown-linux-gnu/native/rust-test-helpers" "-Clinker=x86_64-linux-gnu-gcc"
stdout: none
--- stderr -------------------------------
error: the feature named `sve` is not valid for this target
##[error]  --> /checkout/tests/ui/simd/scalable/copy-clone.rs:11:18
   |
LL | #[target_feature(enable = "sve")]
   |                  ^^^^^^^^^^^^^^ `sve` is not valid for this target

error: the feature named `sve` is not valid for this target
##[error]  --> /checkout/tests/ui/simd/scalable/copy-clone.rs:20:18
   |
LL | #[target_feature(enable = "sve")]
   |                  ^^^^^^^^^^^^^^ `sve` is not valid for this target

error: the feature named `sve` is not valid for this target
##[error]  --> /checkout/tests/ui/simd/scalable/copy-clone.rs:23:18
   |
LL | #[target_feature(enable = "sve")]
   |                  ^^^^^^^^^^^^^^ `sve` is not valid for this target

error: aborting due to 3 previous errors
------------------------------------------


---- [ui] tests/ui/simd/scalable/require-target-feature.rs stdout ----

error: `build-fail` test is required to pass check build, but check build failed
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/simd/scalable/require-target-feature.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=i686-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/simd/scalable/require-target-feature" "-A" "unused" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/i686-unknown-linux-gnu/native/rust-test-helpers" "-Clinker=x86_64-linux-gnu-gcc" "--crate-type=lib"
stdout: none
--- stderr -------------------------------
error: the feature named `sve` is not valid for this target
##[error]  --> /checkout/tests/ui/simd/scalable/require-target-feature.rs:19:18
   |
LL | #[target_feature(enable = "sve")]
   |                  ^^^^^^^^^^^^^^ `sve` is not valid for this target

error: the feature named `sve` is not valid for this target
##[error]  --> /checkout/tests/ui/simd/scalable/require-target-feature.rs:31:18
   |
LL | #[target_feature(enable = "sve")]
   |                  ^^^^^^^^^^^^^^ `sve` is not valid for this target

error: the feature named `sve` is not valid for this target
##[error]  --> /checkout/tests/ui/simd/scalable/require-target-feature.rs:34:18
   |
LL | #[target_feature(enable = "sve")]
   |                  ^^^^^^^^^^^^^^ `sve` is not valid for this target

error: aborting due to 3 previous errors
------------------------------------------


```

</details>

workingjubilee reviewed (changes requested):
I do not intend to have `repr(simd)` survive this year, so I do not think this should be added.

workingjubilee commented:
Likewise, scalable vectors do not really have a diversity in their representation. They either go in the scalable vector registers or they don't. They are one type, even if it is parameterized by the element count (and possibly the CPU's Matrix state, idk).

davidtwco commented:
> I intend to replace it with an approach based on lang items for a variety of reasons, one of them being that to start with, the `repr(simd)` "attribute" does not compose and should not.

I think this discussion is better for the RFC - rust-lang/rfcs#3838 - rather than the implementation.

With that said, I definitely don't think this should to be blocked by your intentions to change `repr(simd)` - extending our existing, [RFC-approved](https://rust-lang.github.io/rfcs/1199-simd-infrastructure.html), SIMD infrastructure is a natural way to go about supporting scalable vectors.  Unless there's an alternative proposal for changing `repr(simd)` out there that has been accepted or is on the cusp of being accepted and I've missed it, working from what we have today is all I can do. 

Nothing in rust-lang/rfcs#3838 should make your plans impossible as it's just more internal infrastructure that could easily be changed to something else with a well-motivated proposal, no different than `repr(simd)`. It'll be your responsibility to update your plans/proposal if `repr(scalable)` makes progress first. It goes the other way too, if you make progress changing `repr(simd)` prior to this making progress then I'll have to update my proposal, that's fine. Unless something I'm proposing completely shuts the door on your plans - which would be worth discussing! - and we decide we'd like to avoid that, then it shouldn't block this.

workingjubilee commented:
1. My understanding has always been that "an RFC approved it" doesn't mean much when it concerns internal-facing details. We've changed things away from RFC-approved implementation details without redoing the RFC. e.g. `#[track_caller]` only barely resembles the original implementation described in the RFC.
2. Specifically, we have removed or heavily modified several parts of that RFC already, and each one was a lot of work.
3. We have had to issue many patches like https://github.com/rust-lang/rust/pull/143833 for fixing how `repr(simd)` is compiled. Yes, we have this existing infrastructure, which is how we know how flawed it is.
4. This is not some random idea I had, it was an issue from some time ago. I am surprised to find out I am the only one who reviews related issues before commencing work on things. https://github.com/rust-lang/rust/issues/63633
5. As someone paid to work on this full-time and also empowered as compiler team lead, I fear you have more ability to push through things than me, especially because for various reasons I take a great deal of care before pushing for things. For that reason, while I am willing to concede the changes here might not obstruct a future change, I do expect to be allowed to review this to make that determination, and hoped for a more collaborative approach instead of possibly fighting to rebase over each other.

davidtwco commented:
Updated this reflecting changes in rust-lang/rfcs#3838. It doesn't implement everything in it yet, and is definitely incomplete, but it's a starting point.

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16909477823/job/47907138539) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/47907138539) [(plain)](https://github.com/rust-lang/rust/commit/6df93d3a1717bcec8d0dba014136eb560d705e9e/checks/47907138539/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] idna_adapter test:false 0.062
error[E0061]: this method takes 3 arguments but 4 arguments were supplied
   --> src/tools/rust-analyzer/crates/hir-ty/src/layout.rs:150:25
    |
150 |     Ok(Arc::new(cx.calc.simd_type(e_ly, e_len, repr_packed, repr_scalable)?))
    |                         ^^^^^^^^^                           ------------- unexpected argument #4 of type `std::option::Option<u32>`
    |
note: method defined here
   --> /checkout/compiler/rustc_abi/src/layout.rs:159:12
    |
159 |     pub fn simd_type<FieldIdx, VariantIdx, F>(
    |            ^^^^^^^^^
help: remove the extra argument
    |
150 -     Ok(Arc::new(cx.calc.simd_type(e_ly, e_len, repr_packed, repr_scalable)?))
150 +     Ok(Arc::new(cx.calc.simd_type(e_ly, e_len, repr_packed)?))
    |

error[E0308]: mismatched types
   --> src/tools/rust-analyzer/crates/hir-ty/src/layout.rs:176:25
    |
172 |                     return layout_of_simd_ty(
    |                            ----------------- arguments to this function are incorrect
...
176 |                         repr.scalable,
    |                         ^^^^^^^^^^^^^ expected `Option<u32>`, found `Option<ScalableElt>`
    |
    = note: expected enum `std::option::Option<u32>`
               found enum `std::option::Option<ScalableElt>`
note: function defined here
   --> src/tools/rust-analyzer/crates/hir-ty/src/layout.rs:122:4
```

</details>

Ruanyx1823 reviewed (commented):
Hello, I'm a student interested in the rust compiler. I'd like to ask how exactly one can write an sve intrinsic. Or how to write the following C language in rust:                                 // #include <arm_sve.h>

// void vec_add(const float *a,const float *b,float *c,int n){
//     for (int i=0; i<n; i+=svcntw()){
//         svbool_t pg = svwhilelt_b32(i,n);
//         svfloat32_t va = svld1(pg,&a[i]);
//         svfloat32_t vb = svld1(pg,&b[i]);
//         svfloat32_t vc = svadd_f32_m(pg,va,vb);
//         svst1(pg,&c[i],vc);
//}
//}

#include <stdio.h>
#include <stdlib.h>
#include <arm_sve.h>

void vec_add(const float* a, const float* b, float* c, int n) {
for (int i = 0;  i < n;  i += svcntw()) {
svbool_t pg = svwhilelt_b32(i, n);
svfloat32_t va = svld1(pg, &a[i]);
svfloat32_t vb = svld1(pg, &b[i]);
svfloat32_t vc = svadd_f32_m(pg, va, vb);
svst1(pg, &c[i], vc);
}
}

void print_array(const char* name, const float* arr, int n) {
printf("%s: [", name);
for (int i = 0;  i < n;  ++i) {
printf("%.1f", arr[i]);
if (i < n - 1) printf(", ");
}
printf("]\n");
}

int main() {
#ifdef __ARM_FEATURE_SVE
printf("SVE is supported! \n");
printf("SVE vector length: %d floats (%d bits)\n",
svcntw(), svcntw() * 32);
#else
printf("SVE not supported.\n");
return 1;
#endif

const int n = 16;
float a[n], b[n], c[n];

for (int i = 0;  i < n;  ++i) {
a[i] = (float)i;
b[i] = (float)(2 * i);
}

printf("\nBefore vector addition:\n");
print_array("A", a, n);
print_array("B", b, n);

vec_add(a, b, c, n);

printf("\nAfter vector addition:\n");
print_array("C = A+B", c, n);

int errors = 0;
for (int i = 0;  i < n;  ++i) {
if (c[i] ! = a[i] + b[i]) {
printf("Error at index %d: expected %.1f, got %.1f\n",
i, a[i] + b[i], c[i]);
errors++;
}
}

if (errors == 0) {
printf("\nAll results correct! \n");
}
else {
printf("\nFound %d errors! \n", errors);
}

return 0;
}

bors commented:
:umbrella: The latest upstream changes (presumably #145085) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

ruanyuxin commented:
Hello, I'm a student interested in rust. I encountered an error when I was testing with Sims.rs. What should I do：
 huawei@DESKTOP-PUT13J1:
 ~/davidtwco/rust$ ./build/x86_64-unknown-linux-gnu/stage2/bin/rustc \
    --target=aarch64-unknown-linux-gnu \
     -C target-feature=+sve,+sve2 \
     --crate-type=lib \
     --sysroot=./build/x86_64-unknown-linux-gnu/stage2 \
     ~/rust-tests/yytest/src/main.rs
error: attributes starting with `rustc` are reserved for use by the `rustc` compiler
--> /home/huawei/rust-tests/yytest/src/main.rs:8:3
|
8 | #[rustc_scalable_vector(4)]
^^^^^^^^^^^^^^^^^^^^^

error: cannot find attribute `rustc_scalable_vector` in this scope
--> /home/huawei/rust-tests/yytest/src/main.rs:8:3
|
8 | #[rustc_scalable_vector(4)]
^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

workingjubilee reviewed (commented):
Some initial questions from my first pass of reading.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/143924/7934bbdf84a6b9a30297caf4f4f38286dedf876a..6cdc8f7f8290bc593a7ae2c0c6bc1a9c74a1725b)*

Amanieu commented:
I don't think having 2 separate types for SVE and SME vectors is going to work. The main issue is that it's absolutely not possible to have scalable locals of 2 different modes in the same function. LLVM doesn't support it, and for good reason since even calculating the stack offsets for locals would involve several SME/SVE mode switches. The problem is that since these types would be runtime-sized, you would be able to smuggle those types into a function in various ways other than parameters and return values, and there's no way you can enforce this in a generic context.

workingjubilee commented:
Hmmmm. Only requiring signature checking does help...

I think I'm not overly attached to a specific direction but I'd like to have the FIXMEs reflect our current inclination, at least.

davidtwco commented on compiler/rustc_abi/src/layout.rs:
```diff
@@ -1501,3 +1477,67 @@ impl<Cx: HasDataLayout> LayoutCalculator<Cx> {
+    Fixed,
+}
+
+fn vector_type_layout<FieldIdx, VariantIdx, F>(
```
Yeah, `simd_type` and now `scalable_vector_type` just forward to this

davidtwco commented on library/core/src/intrinsics/simd.rs:
```diff
@@ -218,6 +218,11 @@ pub const unsafe fn simd_cast<T, U>(x: T) -> U;
+/// Replacement for `transmute`, specifically for use with scalable SIMD types.
+#[rustc_intrinsic]
+#[rustc_nounwind]
+pub unsafe fn simd_reinterpret<Src, Dst>(src: Src) -> Dst;
```
This was a leftover from when these types weren't `Sized`, I've changed it to `transmute_unchecked` in the test that uses it, and we'll need to update the stdarch intrinsic generator to use it too once we start generating the intrinsics.

davidtwco commented on compiler/rustc_mir_transform/src/coroutine.rs:
```diff
@@ -1891,6 +1891,11 @@ fn check_must_not_suspend_ty<'tcx>(
+            tcx.dcx()
+                .span_err(data.source_span, "scalable vectors cannot be held over await points");
+            true
+        }
```
Exactly yeah, added a comment

workingjubilee commented:
It's also unclear whether we will ever be able to enable the compiler to have more ambiguously-sized argument and return types in the future. I think we eventually will want to explicitly ban what we want to ban rather than relying on some other element of the language "doing it for us". If nothing else, for the sake of producing a better error message, at least. That doesn't have to be addressed here, obviously.

workingjubilee commented on compiler/rustc_target/src/callconv/x86_win64.rs:
```diff
@@ -25,6 +25,7 @@ where
                 // FIXME(eddyb) there should be a size cap here
                 // (probably what clang calls "illegal vectors").
             }
+            BackendRepr::ScalableVector { .. } => unreachable!("scalable vectors are unsupported"),
```
I can guess why you used `panic` in one case and `unreachable` in the other, but I'd prefer it was consistent in either direction.
```suggestion
            BackendRepr::ScalableVector { .. } => panic!("scalable vectors are unsupported"),
```

workingjubilee commented on compiler/rustc_abi/src/lib.rs:
```diff
@@ -135,6 +137,19 @@ impl IntegerType {
+)]
+pub enum ScalableElt {
+    /// `N` in `rustc_scalable_vector(N)` - the element count of the scalable vector
+    ElementCount(u128),
```
This can just be u64. We will only parse `0..=u64::MAX` and we will reject with an error things larger than `2.pow(16)`.

workingjubilee commented on compiler/rustc_codegen_ssa/src/mir/debuginfo.rs:
```diff
@@ -408,6 +408,49 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
+                // the `<vscale x 4 x i32>` never spills because that'll cause errors during
+                // instruction selection. Spilling to the stack to create debuginfo for these
+                // intermediate values must be avoided and won't degrade the debugging experience
+                // anyway.
```
...what a journey.

workingjubilee commented on compiler/rustc_hir_analysis/src/check/wfcheck.rs:
```diff
@@ -1037,7 +1037,21 @@ fn check_type_defn<'tcx>(
+                if matches!(ty.kind(), ty::Adt(def, _) if def.repr().scalable())
+                    && !matches!(adt_def.repr().scalable, Some(ScalableElt::Container))
+                {
+                    // Scalable vectors can only be fields of structs if the type has an
```
```suggestion
                    // Scalable vectors can only be fields of structs if the type has a
```

workingjubilee commented:
That last scan-through should be all, unless something is revealed (made apparent?) by the next iteration, so

@rustbot author

davidtwco committed:
abi: `Display` bound on `TyAbiInterface`

The `fmt::Debug` impl for `TyAndLayout<'a, Ty>'` requires `fmt::Display`
on the `Ty` parameter. In `ArgAbi`, `TyAndLayout`'s Ty` is instantiated
with a parameter that implements `TyAbiInterface`. `TyAbiInterface`
only required `fmt::Debug` be implemented on `Self`, not `fmt::Display`,
which meant that it wasn't actually possible to debug print `ArgAbi`.

davidtwco commented on compiler/rustc_target/src/callconv/x86.rs:
```diff
@@ -103,6 +103,9 @@ where
                         false
                     }
+                    BackendRepr::ScalableVector { .. } => {
+                        unreachable!("scalable vectors are unsupported")
```
Fixed

workingjubilee commented:
Seems good to me, r=me with rebase.

rust-timer commented:
Finished benchmarking commit (95a27adcf907bcc1046602c3746dad8cfdf3bf3f): [comparison URL](https://perf.rust-lang.org/compare.html?start=61cc47e367d7be91e13bcd01e4e96e0e233d4f6d&end=95a27adcf907bcc1046602c3746dad8cfdf3bf3f&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=61cc47e367d7be91e13bcd01e4e96e0e233d4f6d&end=95a27adcf907bcc1046602c3746dad8cfdf3bf3f&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.1%  | [0.0%, 0.1%]   | 3     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.6% | [-0.6%, -0.6%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=61cc47e367d7be91e13bcd01e4e96e0e233d4f6d&end=95a27adcf907bcc1046602c3746dad8cfdf3bf3f&stat=max-rss)
<details>
<summary>Results (secondary 1.2%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.3%  | [1.6%, 4.4%]   | 3     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.8% | [-3.0%, -0.6%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=61cc47e367d7be91e13bcd01e4e96e0e233d4f6d&end=95a27adcf907bcc1046602c3746dad8cfdf3bf3f&stat=cycles:u)
<details>
<summary>Results (secondary -1.7%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.8%  | [3.8%, 3.8%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.5% | [-4.9%, -2.4%] | 3     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=61cc47e367d7be91e13bcd01e4e96e0e233d4f6d&end=95a27adcf907bcc1046602c3746dad8cfdf3bf3f&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.1%, secondary 0.0%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1% | [0.0%, 0.2%] | 54    |
| Regressions ❌ <br /> (secondary)  | 0.0% | [0.0%, 0.1%] | 45    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.1% | [0.0%, 0.2%] | 54    |
</details>

**Bootstrap**: 478.546s -> 479.485s (0.20%)
**Artifact size**: 390.29 MiB -> 390.32 MiB (0.01%)
